fb: cfb: Support color display

soburi opened the pull request:
Add color display support for CFB.

The concept of "Invert" depended on monochrome displays.
I reimplemented it with foreground color and background color concepts that are newly introduced with color support.

The CFB is no longer a "(Monochrome) Character Frame Buffer" but 
the concept and usage of the API remain the same.
Since I would like to maintain the CFB abbreviation, I am considering renaming it "Compact Frame Buffer."

---------------------

This PR is part of a series of CFB enhancements (third of four https://github.com/zephyrproject-rtos/zephyr/issues/72177). 
I would appreciate it if you could check the [#67881](https://github.com/zephyrproject-rtos/zephyr/pull/68260)  part first.

MaureenHelm commented:
@jfischer-no @danieldegrasse will you please take a look?

danieldegrasse reviewed (commented):
Just some initial comments- but I am still confused about the goal of this rework. First off, the subsystem itself is called [Monochrome Character Framebuffer](https://docs.zephyrproject.org/latest/hardware/peripherals/display/index.html#monochrome-character-framebuffer). The subsystem is explicitly:
- monochrome
- intended for font rendering

If we are going to extend CFB to be what I would consider a small graphics framework, I would like to see the following:
- Documentation updates- the current CFB naming is confusing
- Additional tests or samples demonstrating usage of this framework, so we can verify rendering functions work as expected

Additionally, I am slightly confused about where a graphics framework would fit versus LVGL. Is the intention of this framework to fit where LVGL can't?

@jfischer-no ping- I'll defer to you on any of these comments, I just want to make sure that if we are extending CFB we keep documentation and samples up to date.

jfischer-no commented on subsys/fb/cfb.c:
```diff
@@ -440,30 +544,95 @@ int cfb_framebuffer_invert(const struct device *dev)
+		return api->write(dev, 0, 0, &desc, fb->buf);
+	}
+
+#ifdef CONFIG_CHARACTER_FRAMEBUFFER_TRANSFER_WITH_PIXEL_FORMAT_CONVERSION
```
```c
if(ENABLED(CFB_TRANSFER_WITH_PIXEL_FORMAT_CONVERSION)) {
	cfb_transfer_with_pixel_conversion(..);
}

```
I do not understand why this code should not always be built, and why it cannot be broken into smaller chunks. As it is now, it is difficult to understand and review.

MaureenHelm commented:
@soburi do you plan to return to this PR?

soburi commented:
> @soburi do you plan to return to this PR?

I'm in bit busy state. I'll late to update it.

soburi committed:
tests: subsys: display: cfb: Added test for font heights non-x8

Add test cases for 1x1 and 11x13 fonts.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb: Remove the restriction for non-x8 heights fonts

The logic introduced in #67881 can also handle fonts with heights
other than integer multiple of 8. I remove the restriction.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb: Supporting multiple display

There are three major changes.

- Introducing `struct cfb_display`
- Changed the argument `struct device *` to `struct cfb_framebuffer *`
- Rename APIs starting with `cfb_framebuffer_*`

Introduces struct cfb_display as a data structure to represent the display.
As a result, it is now possible to create multiple pieces of data managed
using a single static variable in previous implementations,
making it possible to handle multiple displays.
The old `struct cfb_framebuffer` remains but is now part of the
`struct cfb_display` data structure.
You can get cfb_framebuffer from cfb_display using the added API
`cfb_display_get_framebuffer`.

All places where a `struct device *` argument was passed to specify the
target will be replaced with a `struct cfb_framebuffer *`.
In line with this, API names starting with `cfb_framebuffer_*`,
which were redundant and inconsistent names, will remove
`_framebuffer` and change to `cfb_*` .

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb: Remove unused argument from font API

`cfb_get_font_size` and `cfb_get_numof_fonts` take a
`const struct driver *` as a first argument, but not use it.

Remove the argument and clean up related sources.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
samples: subsys: display: Update CFB samples for supporting color

I have removed the process to set the display format to monochrome.
Also, I added board settings conf files for native_sim and native_sim_64
to change the initial value of the SDL display driver's pixel format.
CFB operates according to the pixel format set here.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb: Remove unused config

CHARACTER_FRAMEBUFFER_SHELL_DRIVER_NAME is not used.
Remove it.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb: Changed library name to Compact framebuffer

CFB was an abbreviation for Character Framebuffer.
However, it currently supports characters and graphic drawing,
so it has become far from actual.

However, since there are only a few API changes, I don't want to
change symbol names, etc.
Therefore, we will keep the abbreviation CFB and change the name
to `Compact Framebuffer`.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb_shell: Update README.rst

Update README to update to current specs.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb: Delete ppt field of cfb_framebuffer struct

The pixel per tile value can be calculated from pixel_format,
so delete the ppt field.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
doc: releases: migration-guide: 3.7: list CFB API change

Add a note about the changes of CFB(Character framebuffer) APIs.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb: Improved behavior when a character not in the font

If a character that does not exist in the font was specified,
it was replaced with a space character, but it will now draw with 0 data.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
test: subsys: display: cfb: Add test for color display

Added a test to check the display on a color display.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb_shell: Add command for color display supporting

Add the `fgcolor` `bgcolor` and `display pixel_format`  to
support color display

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb_shell: Support display selecting

Add `display select` command to select display.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb: Support color display

Added support for color displays.

Along with color support, we are introducing the concept of foreground
and background colors. Drawing fonts and lines is done with the
foreground color, and clear operation fills with the background color.
With the introduction of drawing colors, we have removed the process of
inverting data before transfer if the inverted flag was enabled
during finalizing.
Replacing the "Invert" operation by swapping the foreground and
background colors.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
fb: cfb: Use signed-integer for API that specifies coordinations

Change the arguments of `cfb_invert_area`, `cfb_print`, and
`cfb_draw_(point|line|rect)`, which used unsigned integers to specify
coordinates to signed integers.

This is because coordinate calculations are usually based on real
numbers, so it is difficult to use unless negative numbers are
accepted.
We clip drawings outside the drawable area on the implementation side.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

cfriedt commented on include/zephyr/display/cfb.h:
```diff
@@ -51,8 +52,72 @@ struct cfb_font {
+	int16_t x;
+	int16_t y;
+};
+
```
For private structures, it would be best to hide them from Doxygen. This requires an `@endcond` after the hidden declarations.

```suggestion
/**
 * @cond INTERNAL_HIDDEN
 */
```

soburi replied:
Thank you for your review.
I fixed the pointing out.
And I also cleaned up the overall doxygen output.

danieldegrasse commented:
@jfischer-no can you revisit comments on this PR?

kartben commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -277,6 +277,65 @@ Networking
+    cfb_print(fb, "Hello!", 0, 0);
+
+
+  Also, please see the sample in ``samples/display/cfb``.
```
path is wrong, and you may use `:zephyr:code-sample:` anyway 
```suggestion
  Also, please see the :zephyr:code-sample:`cfb-sample`.
```

kartben commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -277,6 +277,65 @@ Networking
+  * :c:func:`cfb_finalize` (renamed from `cfb_framebuffer_finalize`)
+  * :c:func:`cfb_set_font` (renamed from `cfb_framebuffer_finalize`)
+
+  :c:func:`cfb_display_alloc` can use to migrate from `cfb_framebuffer_init`.
```
not clear what this means?

kartben commented on doc/releases/migration-guide-3.7.rst:
would be nice to have release note entry as well, but this can come later too.

kartben reviewed (changes requested):
some comments re: docs / migration guide
